### PR TITLE
add appengine/bigquery sample

### PR DESCRIPTION
**Pull request by campoy**
FYI @broady @markmandel

### CODE REVIEW

File: appengine/bigquery/app.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Copyright 2015 Google Inc. All rights reserved.
+# Use of this source code is governed by the Apache 2.0
+# license that can be found in the LICENSE file.
+application: gae-bq-sample
```
Comment (broady) [inline appengine/bigquery/app.yaml]: Is this still required for App Engine? I think you can specify it on the command line.

File: appengine/bigquery/README.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
+# Accessing BigQuery from App Engine
```
Comment (broady) [inline appengine/bigquery/README.md]: Add a short section with deployment instructions.

File: appengine/bigquery/README.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+A Google Cloud Platform customer asked me today how to list all the BigQuery
+projects that you own from a Google App Engine app.
```
Comment (broady) [inline appengine/bigquery/README.md]: This reads like a blog post. You may want to add your name above, or re-word so it's more like a normal README.

File: appengine/bigquery/README.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+by the App Engine app. Finally, the list will be printed to the output `w`.
+
+In order to register the handler so it is executed on every HTTP request we add
+an `init` func.
+
+```go
+func init() {
+
+    // all requests are handled by handler.
+    http.HandleFunc(“/”, handle)
```
Comment (broady) [inline appengine/bigquery/README.md]: smart quotes

File: appengine/bigquery/app.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	"google.golang.org/appengine"
+	"google.golang.org/appengine/urlfetch"
+)
+
+func init() {
+	// all requests are handled by handler.
+	http.HandleFunc("/", handle)
+}
+
+func handle(w http.ResponseWriter, r *http.Request) {
```
Comment (broady) [inline appengine/bigquery/app.go]:
```
if r.Path != "/" {
// 404
}

File: appengine/bigquery/app.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/bigquery/v2"
+	"google.golang.org/appengine"
+	"google.golang.org/appengine/urlfetch"
+)
+
+func init() {
+	// all requests are handled by handler.
```
Comment (broady) [inline appengine/bigquery/app.go]: remove this comment

File: appengine/bigquery/app.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// print it to the output.
+	fmt.Fprintln(w, strings.Join(names, "\n"))
+}
+
+// projects returns a list with the names of all the Big Query projects visible
+// with the given context.
```
Comment (broady) [inline appengine/bigquery/app.go]: it seems like it'll list all the bigquery projects in the same project as the running App Engine app.

File: appengine/bigquery/app.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+		Transport: &oauth2.Transport{
+			Source: google.AppEngineTokenSource(c, bigquery.BigqueryScope),
+			Base:   &urlfetch.Transport{Context: c},
+		},
+	}
+
+	// create the BigQuery service.
+	bq, err := bigquery.New(client)
+	if err != nil {
+		return nil, fmt.Errorf("create service: %v", err)
```
Comment (broady) [inline appengine/bigquery/app.go]: `could not create service:`

File: appengine/bigquery/app.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	// list the projects.
+	list, err := bq.Projects.List().Do()
+	if err != nil {
+		return nil, fmt.Errorf("list projects: %v", err)
```
Comment (broady) [inline appengine/bigquery/app.go]: `could not list projects:`

File: appengine/bigquery/app.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	// prepare the list of names.
+	var names []string
+	for _, p := range list.Projects {
+		names = append(names, p.FriendlyName)
+	}
+	return names, nil
```
Comment (broady) [inline appengine/bigquery/app.go]: maybe

```
if len(list.Projects) == 0 {
  return nil, errors.New("no bigquery projects")
}

File: appengine/bigquery/app.yaml (hunk 1 above)
Comment (campoy) [inline appengine/bigquery/app.yaml]: done

File: appengine/bigquery/README.md (hunk 3 above)
Comment (campoy) [inline appengine/bigquery/README.md]: I removed this so it points to the medium post instead.

File: appengine/bigquery/README.md (hunk 2 above)
Comment (campoy) [inline appengine/bigquery/README.md]: Done

File: appengine/bigquery/README.md (hunk 4 above)
Comment (campoy) [inline appengine/bigquery/README.md]: done

File: appengine/bigquery/app.go (hunk 5 above)
Comment (campoy) [inline appengine/bigquery/app.go]: I think it's fine like this, ANY url will be handled by this.

File: appengine/bigquery/app.go (hunk 6 above)
Comment (campoy) [inline appengine/bigquery/app.go]: see comment above.

File: appengine/bigquery/app.go (hunk 9 above)
Comment (campoy) [inline appengine/bigquery/app.go]: done

File: appengine/bigquery/app.go (hunk 10 above)
Comment (campoy) [inline appengine/bigquery/app.go]: done

File: appengine/bigquery/app.go (hunk 5 above)
Comment (broady) [inline appengine/bigquery/app.go]: Including `favicon.ico`. It's wasteful to basically double up on requests.

File: appengine/bigquery/app.go (hunk 6 above)
Comment (broady) [inline appengine/bigquery/app.go]: I mean, it's not a very useful comment.

File: appengine/bigquery/app.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+	if len(names) == 0 {
+		fmt.Fprintf(w, "no datasets visible")
+	} else {
+		fmt.Fprintf(w, "datasets:\n\t"+strings.Join(names, "\n\t"))
+	}
+}
+
+// datasets returns a list with the ids of all the Big Query datasets visible
+// with the given context.
+func datasets(c context.Context) ([]string, error) {
```
Comment (broady) [inline appengine/bigquery/app.go]: `c` --> `ctx`

File: appengine/bigquery/app.yaml [hunk 12]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2015 Google Inc. All rights reserved.
+# Use of this source code is governed by the Apache 2.0
+# license that can be found in the LICENSE file.
```
Comment (broady) [inline appengine/bigquery/app.yaml]: no need for copyright header on app.yaml

File: appengine/bigquery/app.go (hunk 5 above)
Comment (campoy) [inline appengine/bigquery/app.go]: done

File: appengine/bigquery/app.go (hunk 6 above)
Comment (campoy) [inline appengine/bigquery/app.go]: done

File: appengine/bigquery/app.go (hunk 8 above)
Comment (campoy) [inline appengine/bigquery/app.go]: done

File: appengine/bigquery/app.go (hunk 11 above)
Comment (campoy) [inline appengine/bigquery/app.go]: done

File: appengine/bigquery/app.yaml (hunk 12 above)
Comment (campoy) [inline appengine/bigquery/app.yaml]: is that new? ok, done

File: appengine/bigquery/app.go [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+func handle(w http.ResponseWriter, r *http.Request) {
+	if r.URL.Path != "/" {
+		http.NotFound(w, r)
+		return
+	}
+
+	// create a new App Engine context from the request.
+	c := appengine.NewContext(r)
```
Comment (broady) [inline appengine/bigquery/app.go]: I think we want `c` -> `ctx`.

(`ctx` because we're using the actual `context` package now.)

Comment (campoy) [inline appengine/bigquery/app.go]: done

**Comment by broady** [comment]
merged e317f76e6c265283308ad1d558bc9b4fa22597cf

**Pull request closed without merging**